File: test/The 2024 ICPC Asia Shenyang Regional Contest/G3.cpp
#include<bits/stdc++.h>
using namespace std;
int n,m;
set<int> xs;
vector<int> vx;
void Answer(double ans){
    int area2 = round(2*ans);
    if(area2%2==0) cout<<"! "<<area2/2<<" 1"<<endl;
    else cout<<"! "<<area2<<" 2"<<endl;
}
vector<double> sumer;
void work1(){
    double len=0,lstlen,ans=0;
    for(int i=1;i<n-1;++i){
        lstlen=len;
        cout<<"? "<<vx[i]<<" 1"<<endl;
        double x,y;cin>>x>>y;len=x/(double)y;
        sumer.push_back((len+lstlen)/2*(vx[i]-vx[i-1]));
    }
    sumer.push_back(len/2*(vx[n-1]-vx[n-2]));
    sort(sumer.begin(),sumer.end());
    for(auto x:sumer) ans+=x;
    Answer(ans);
}
void work2(){
    double ans=0;
    for(int i=1;i<m;++i){
        int midx2 = vx[i]+vx[i-1];
        if(midx2%2==0) cout<<"? "<<midx2/2<<" 1"<<endl;
        else cout<<"? "<<midx2<<" 2"<<endl;
        double x,y;cin>>x>>y;double len=x/(double)y;
        sumer.push_back(len*(vx[i]-vx[i-1]));
    }
    sort(sumer.begin(),sumer.end());
    for(auto x:sumer) ans+=x;
    Answer(ans);
}
void work3(){
    int x[10],y[10];
    for(int i=1;i<=3;++i){
        cin>>x[i]>>y[i];
    }
    x[1]-=x[3]; y[1]-=y[3];
    x[2]-=x[3]; y[2]-=y[3];
    int area = abs(x[1]*y[2]-y[1]*x[2]);
    Answer(area/2.0);
}
void work(){
    cin>>n; xs.clear(); vx.clear(); sumer.clear();
    if(n==3) {work3();return;}
    for(int i=1;i<=n;++i){
        int x,y;
        cin>>x>>y;
        xs.insert(x);
    }
    m=(int)xs.size();
    for(auto x:xs) vx.push_back(x);
    if(m==n) work1();
    else work2();
}
signed main(){
    int T;
    cin>>T;
    while(T--) work();
}

File: test/The 2024 ICPC Asia Shenyang Regional Contest/G3.py
import sys

def gcd(x,y):
    if y==0:
        while True:
            pass
    if x==0:
        return y
    while x!=0:
        y %= x
        t=x
        x=y
        y=t
    return y
def add(x,y):
    p = x[0]*y[1]+x[1]*y[0]
    q = x[1]*y[1]
    g = gcd(p,q)
    return (p//g,q//g)
def mul(x,y):
    p = x[0]*y[0]
    q = x[1]*y[1]
    g = gcd(p,q)
    return (p//g,q//g)

def Answer(ans):
    print("!",ans[0],ans[1])
    sys.stdout.flush()

def work1():
    global n
    global vx
    len=(0,1)
    ans=(0,1)
    for i in range(1,n-1):
        lstlen=len
        print("?",vx[i],1)
        sys.stdout.flush()
        instr = input().split()
        len = (int(instr[0]),int(instr[1]))
        ans = add(ans,mul(mul(add(len,lstlen),(1,2)),(vx[i]-vx[i-1],1)))
    ans = add(ans,mul(mul(len,(1,2)),(vx[n-1]-vx[n-2],1)))
    Answer(ans)

def work2():
    global n
    global vx
    ans=(0,1)
    for i in range(1,m):
        midx2 = vx[i]+vx[i-1]
        if(midx2%2==0):
            print("?",midx2//2,"1")
        else:
            print("?",midx2,"2")
        sys.stdout.flush()
        instr = input().split()
        len = (int(instr[0]),int(instr[1]))
        ans = add(ans,mul(len,(vx[i]-vx[i-1],1)))
    Answer(ans)

def work():
    global n
    global m
    global xs
    global vx
    n=int(input())
    xs={}
    vx=[]
    for i in range(1,n+1):
        instr=input().split()
        x=int(instr[0])
        xs[x]=True
    for i in xs:
        vx.append(i)
    m = len(vx)
    if(m==n):
        work1()
    else:
        work2()


T = int(input())
while T>0:
    T-=1
    work()